python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,558 @@
+        or other, but not both.
+
+        Parameters
+        ----------
+        other : iterable over `DimensionElement` or `str`.
+            The other set from which to draw potential result elements.
+
+        Returns
+        -------
+        result : `DimensionSet`
+            A new set containing elements ``self`` or ``other``, but not both.
+        """
+        names = self.names ^ toNameSet(other).names
+        return DimensionSet(self.universe, names)
+
+    def difference(self, other):
+        """Return a new set containing all elements that are in ``self``
+        but not other.
+
+        Parameters
+        ----------
+        other : iterable over `DimensionElement` or `str`.
+            The other set containing elements that should not be included
+            in the result.
+
+        Returns
+        -------
+        result : `DimensionSet`
+            A new set containing elements in ``self`` but not ``other``.
+        """
+        names = self.names - toNameSet(other).names
+        return DimensionSet(self.universe, names)
+
+    # Operators that return sets are only enabled for DimensionSet operands on
+    # both sides, to avoid confusion about return types.
+
+    def __or__(self, other):
+        if isinstance(other, DimensionSet):
+            return self.union(other)
+        return NotImplemented
+
+    def __and__(self, other):
+        if isinstance(other, DimensionSet):
+            return self.intersection(other)
+        return NotImplemented
+
+    def __xor__(self, other):
+        if isinstance(other, DimensionSet):
+            return self.symmetric_difference(other)
+        return NotImplemented
+
+    def __sub__(self, other):
+        if isinstance(other, DimensionSet):
+            return self.difference(other)
+        return NotImplemented
+
+    def findIf(self, predicate, default=None):
+        """Return the element in ``self`` that matches the given predicate.
+
+        Parameters
+        ----------
+        predicate : callable
+            Callable that takes a single `DimensionElement` argument and
+            returns a `bool`, indicating whether the given value should be
+            returned.
+        default : `DimensionElement`, optional
+            Object to return if no matching elements are found.
+
+        Returns
+        -------
+        matching : `DimensionElement`
+            Element matching the given predicate.
+
+        Raises
+        ------
+        ValueError
+            Raised if multiple elements match the given predicate.
+        """
+        t = tuple(element for element in self if predicate(element))
+        if len(t) > 1:
+            raise ValueError(f"Multiple matches: {t}")
+        elif len(t) == 0:
+            return default
+        return t[0]
+
+
+class DimensionNameSet:
+    r"""An incomplete, name-only stand-in for `DimensionSet` or
+    `DimensionGraph`.
+
+    Because true `DimensionSet`\s and `DimensionGraph`\s cannot be constructed
```

Comment:
Move these paragraphs to a [Notes](
https://developer.lsst.io/python/numpydoc.html#notes) section.